src/main/java/co/cask/spanner/source/SpannerRecordReader.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.spanner.source;
+
+import com.google.cloud.spanner.ResultSet;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Spanner record reader - updates result set during iteration
+ */
+public class SpannerRecordReader extends RecordReader<NullWritable, ResultSet> {
+  private static final Logger LOG = LoggerFactory.getLogger(SpannerRecordReader.class);
+  PartitionInputSplit partitionInputSplit;
+  ResultSet resultSet;
+  boolean isFinished;
+
+  @Override
+  public void initialize(InputSplit inputSplit,
+                         TaskAttemptContext taskAttemptContext) throws IOException, InterruptedException {
+    partitionInputSplit = (PartitionInputSplit) inputSplit;
+    try {
+      resultSet = partitionInputSplit.getResultSet();
+      isFinished = false;
+    } catch (Exception e) {
+      throw new IOException("Exception while trying to execute query to get result set ", e);
+    }
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException, InterruptedException {
+    if (resultSet != null && resultSet.next()) {
+      return true;
+    }
+    return false;
+  }
+
+  @Override
+  public NullWritable getCurrentKey() throws IOException, InterruptedException {
+    return NullWritable.get();
+  }
+
+  @Override
+  public ResultSet getCurrentValue() throws IOException, InterruptedException {
+    return resultSet;
+  }
+
+  @Override
+  public float getProgress() throws IOException, InterruptedException {
+    return isFinished ? 1.0f : 0.0f;
```

Comment:
I assume this means ResultSet doesn't tell you how many results there are?

Reply:
yes.

Reply:
Do you know if this ever gets called after close(). Seems like it would be a weird contract. If not, you can make this always return 0.0f and remove the isFinished flag.